### ISSUE TITLE
ParamSpec in parametrizable decorators doesn't infer usable types

### ISSUE THREAD

**Issue by Agalin** [microsoft/pyright#1907]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
I'm trying to write type hints for a decorator that takes a callback for which all call args are passed (e.g. logging/tracing callback) but it seems to be impossible to type correctly due to order of *ParamSpec* binding.

For function-based decorator it seems to force it to use callback's signature. Not only all type hints from decorated function are lost but generic callbacks are basically unusable.

For class-based decorator it seems to consider ParamSpecs independently even when *Generic[P]* is used. Looking at examples in PEP612 it probably should behave like function-based one in the case with *Generic[P]*. Instead we get type hints from the decorated function but no validation between the callback and the decorated function at all.

**To Reproduce**
See code.
Missing error - might lead to a runtime exception.
Incorrect error - will work at run time.

**Expected behavior**
See comments in code.

**Screenshots or Code**
Inheriting from *Generic[P]* is commented out due to incompatibility in *ParamSpec* backport but it doesn't affect anything anyway.
```python
from functools import wraps
from typing import TypeVar, Callable, Generic
from typing_extensions import ParamSpec

P = ParamSpec("P")
R = TypeVar("R")


# no difference for def decorate(cb: Callable[P, None]) -> Callable[[Callable[P, R]], Callable[P, R]]:
def decorate(cb: Callable[P, None]):
    def decorator(f: Callable[P, R]) -> Callable[P, R]:
        @wraps(f)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            cb(*args, **kwargs)
            return f(*args, **kwargs)

        return wrapper

    return decorator


# no difference for class cls_decorate(Generic[P]):
class cls_decorate:
    def __init__(self, cb: Callable[P, None]):
        self.cb = cb

    def __call__(self, f: Callable[P, R]) -> Callable[P, R]:
        @wraps(f)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            self.cb(*args, **kwargs)
            return f(*args, **kwargs)

        return wrapper


@decorate(lambda bar, *, baz: print("cb"))
def foo(bar: int, *, baz: str):
    print(f"foo: {bar}, {baz}")


@decorate(lambda *args, **kwargs: print("cb"))  # incorrect error
def bar(baz: int, *, foobar: str):
    print(f"bar: {baz}, {foobar}")


@decorate(lambda foo, *, baz: print("cb"))  # correct error
def baz(foo: int):
    print(f"baz: {foo}")


@cls_decorate(lambda bar, *, baz: print("cb"))
def cls_foo(bar: int, *, baz: str):
    print(f"cls_foo: {bar}, {baz}")


@cls_decorate(lambda *args, **kwargs: print("cb"))
def cls_bar(baz: int, *, foobar: str):
    print(f"cls_bar: {baz}, {foobar}")


@cls_decorate(lambda foo, *, baz: print("cb"))  # missing error
def cls_baz(foo: int):
    print(f"cls_baz: {foo}")


foo(1, baz="str")  # foo: (bar: Unknown, *, baz: Unknown) -> None - types missing
bar(2, foobar="str")  # bar: (*args: Unknown, **kwargs: Unknown) -> None - everything missing
cls_foo(1, baz="str")  # cls_foo: (bar: int, *, baz: str) -> None - signature correct
cls_bar(2, foobar="str")  # cls_bar: (baz: int, *, foobar: str) -> None - signature correct

foo("str", baz="str")  # missing error
bar("str", foobar="str")  # missing error
cls_baz(1)  # will raise a TypeError about missing baz at runtime.
```

**VS Code extension or command-line**
Both Pylance 2021.5.3 in VSCode and Pyright 1.1.143 in shell.

**Additional context**
Pyright output (paths removed)
```
demo.py:42:2 - error: Argument of type "(baz: int, *, foobar: str) -> None" cannot be assigned to parameter "f" of type "(*args: Unknown, **kwargs: Unknown) -> R@decorator" in function "decorator"
    Type "(baz: int, *, foobar: str) -> None" cannot be assigned to type "(*args: Unknown, **kwargs: Unknown) -> R@decorator"
      Parameter *args has no corresponding parameter
      Parameter *kwargs has no corresponding parameter (reportGeneralTypeIssues)
demo.py:47:2 - error: Argument of type "(foo: int) -> None" cannot be assigned to parameter "f" of type "(foo: Unknown, *, baz: Unknown) -> R@decorator" in function "decorator"
    Type "(foo: int) -> None" cannot be assigned to type "(foo: Unknown, *, baz: Unknown) -> R@decorator"
      Keyword parameter "baz" is missing in source (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos
Completed in 0.776sec
```

Edit. I hope I didn't miss anything, needed to rewrite issue from scratch due to some Github error. 😢

**Comment by erictraut** [comment]
Thanks for writing this issue... twice.

Let me try to explain what's going on here. In most respects, ParamSpec works like any other TypeVar in Python. Each TypeVar has a defined scope consisting of a generic class, a generic function, or a generic type alias. A TypeVar is "resolved" — that is, it is assigned an actual type — when you call a constructor (for a generic class) or call a function (for a generic function). The process is called "constraint resolution" because it attempts to "solve" each of the type variables in a way that satisfies all of their constraints.

With that explanation in mind, let's look at what's going on in your code above.

First, I'm going to simplify your decorator function `decorate` to remove its implementation. From a type checker's perspective, the signature is what's important when evaluating a call expression.

```python
def decorate(cb: Callable[P, None]) -> Callable[[Callable[P, R]], Callable[P, R]]:
    ...
```

Next, I'm going to define a non-lambda function that is the equivalent of your first lambda, once again with the implementation omitted. This will help make it clearer what's going on. Note that I've left the parameters `*args` and `**kwargs` unannotated, reflecting the fact that Python provides no way to annotate parameters in lambdas.

```python
def lambda_equivalent(*args, **kwargs) -> None:
    ...
```

Now, let's consider the following expression:
```python
d = decorate(lambda_equivalent)
```

What is the type of `d`? Pyright is going to use the signature of `decorate` and the argument `lambda_equivalent` to resolve `P` and `R`. The resulting type is: 
```
(_p0: (*args: Unknown, **kwargs: Unknown) -> R@decorate) -> (*args: Unknown, **kwargs: Unknown) -> R@decorate
```

Note that `P` has been resolved and expanded to reflect the input signature of `lambda_equivalent`. The `R` TypeVar wasn't matched by any input arguments, so it remains unresolved.

Now, what happens if we apply the resulting decorator function `d`?

```python
def bar(baz: int, *, foobar: str): ...

d(bar) # Error
```

This results in an error because the first parameter to `d` must accept `*args` and `**kwargs`, which `bar` does not. So I think this error is correct.

Let's take a look at your class-based decorator. The problem here is that you have declared a non-generic class with two generic methods that both use `P` and `R`. The `P` used in `__init__` has no relation to the `P` used in `__call__`. I think your intent was the declare a generic class, like this:
```python
class cls_decorate(Generic[P, R]):
    def __init__(self, cb: Callable[P, None]):
        ...

    def __call__(self, f: Callable[P, R]) -> Callable[P, R]:
        ...
```

With this change, you will see the same result for `decorate` and `cls_decorate` because they are now equivalent.

I realize that type variable scoping is confusing in Python. There has been recent talk among the Python type community to define a better syntax that would help alleviate some of this confusion, but nothing concrete has yet come of those discussions.

In summary, I think pyright is doing the right thing in this case. (I did notice one small bug in the error message. It says "Parameter *kwargs..." when it should read "Parameter **kwargs...". I've fixed that bug.)

**Comment by Agalin** [comment]
Ah. Exactly what I've feared: "expected state" is unusable in this case. 😞 

But why must it be `Generic[P, R]`? `__init__` doesn't use `R` at all so for `cls_decorate(Generic[P])` I'd expect `__call__` to have `P` resolved and `R` still unbound. And yet it seems to completely ignore `Generic[P]` in this case.

As in my initial example (commented out part):
```python
class cls_decorate(Generic[P]):
    def __init__(self, cb: Callable[P, None]):
        ...

    def __call__(self, f: Callable[P, R]) -> Callable[P, R]:
        ...
```
`P` seems not to be shared at all - with quite a strange behavior in comparison to how other generics work.

```python
def lambda_equivalent(*args, **kwargs) -> None:
    ...

d = cls_decorate(lambda_equivalent) # d: cls_decorate[P@cls_decorate]
                                    # d.__call__: (f: Callable[P, R@__call__]) -> Callable[P, R@__call__]
def cls_bar(baz: int, *, foobar: str):
    print(f"cls_bar: {baz}, {foobar}")

result = d(cls_bar) # result: (baz: int, *, foobar: str) -> None
```
So `cls_decorate[P@cls_decorate]` has `P` scoped to the whole class and `R` scoped to `__call__` method (as expected). But `P` in the `__call__` method is a completely different `P` (even though it doesn't have `@__call__` even for non-generic class I'd assume that it's also scoped to `__call__`). Why isn't it `P@cls_decorate`? Why doesn't it show `@__call__` when `R` does while it would make more sense to show it for `P` (that has 2 meanings in different scopes of this class) than for `R` that's only used in one place? Is it assumed that `P` shares `@__call__` part with `R`? But it doesn't have `@__init__` either for that method so it doesn't seem to be the case .

For the full generic (`class cls_decorate(Generic[P, R])`) it doesn't mark `P` in the method signature as the same as in the generic as well - while type checking works just as you said it should - shows errors the same way as for function.
But! It doesn't seem to resolve `R` instead just assume it's unknown during the initialization.

```python
def decorate(cb: Callable[P, None]) -> Callable[[Callable[P, R]], Callable[P, R]]:
    ...

class cls_decorate(Generic[P, R]):
    def __init__(self, cb: Callable[P, None]):
        ...

    def __call__(self, f: Callable[P, R]) -> Callable[P, R]:
        ...

def lambda_equivalent(*args, **kwargs) -> None:
    ...

d = decorate(lambda_equivalent) # d: (_p0: (*args: Unknown, **kwargs: Unknown) -> R@decorate) -> (*args: Unknown, **kwargs: Unknown) -> R@decorate
d2 = cls_decorate(lambda_equivalent) # d2: cls_decorate[(*args: Unknown, **kwargs: Unknown), Unknown]

```
It means that when using class decorator even the return value of the final decorated function is lost while it's preserved for the function-based one.
```python
def func(*args, **kwargs) -> str:
    ...

d_func = d(func) # -> str
d2_func = d2(func) # -> Unknown
```

**Comment by erictraut** [comment]
You are correct that you could declare the `cls_decorate` class with only one generic type parameter `P`.

>`P` seems not to be shared at all - with quite a strange behavior in comparison to how other generics work.

You found a bug here, but it's not in the type checker itself. Rather, it's a bug in the code that converts the type to text. It was omitting the `@scope` portion in this code path. I've fixed this problem, since it's inconsistent and misleading, but this issue doesn't have any impact on the core type checking logic.

You've also found a bug in the case where `P` is not specialized in the call to the class constructor when `P` is the only type variable for the class. I've fixed that bug as well.

Thanks for reporting these issues. You're using ParamSpec in creative ways that I didn't anticipate when writing unit tests for this feature.

**Comment by Agalin** [comment]
Glad I could help at least a bit. 😄 Really unfortunate that my main use case cannot be supported. Maybe in Python 3.14... as I don't think that many people will be affected before wide 3.10 adoption - it's quite specific use case to pass callbacks this way in Python.
Anyway I think we can close this issue then.

**Comment by erictraut** [comment]
These bugs are addressed in pyright 1.1.144, which I just published. It will also be in the next release of pylance.
